File: 351002/Lashchenka/publisher/src/main/java/by/bsuir/distcomp/service/TweetService.java
package by.bsuir.distcomp.service;

import by.bsuir.distcomp.dto.request.TweetRequestTo;
import by.bsuir.distcomp.dto.response.TweetResponseTo;
import by.bsuir.distcomp.entity.Tweet;
import by.bsuir.distcomp.exception.DuplicateException;
import by.bsuir.distcomp.exception.ResourceNotFoundException;
import by.bsuir.distcomp.mapper.TweetMapper;
import by.bsuir.distcomp.repository.EditorRepository;
import by.bsuir.distcomp.repository.TweetRepository;
import org.springframework.cache.annotation.CacheEvict;
import org.springframework.cache.annotation.CachePut;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.cache.annotation.Caching;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;

@Service
@Transactional
public class TweetService {

    private final TweetRepository tweetRepository;
    private final EditorRepository editorRepository;
    private final TweetMapper tweetMapper;

    public TweetService(TweetRepository tweetRepository,
                        EditorRepository editorRepository,
                        TweetMapper tweetMapper) {
        this.tweetRepository = tweetRepository;
        this.editorRepository = editorRepository;
        this.tweetMapper = tweetMapper;
    }

    @Caching(put = @CachePut(value = "tweets", key = "#result.id"),
            evict = @CacheEvict(value = "tweets", key = "'all'"))
    public TweetResponseTo create(TweetRequestTo dto) {
        if (!editorRepository.existsById(dto.getEditorId())) {
            throw new ResourceNotFoundException("Editor with id " + dto.getEditorId() + " not found", 40404);
        }
        if (tweetRepository.existsByTitle(dto.getTitle())) {
            throw new DuplicateException("Tweet with title '" + dto.getTitle() + "' already exists", 40303);
        }
        Tweet entity = tweetMapper.toEntity(dto);
        entity.setCreated(LocalDateTime.now());
        entity.setModified(LocalDateTime.now());
        Tweet saved = tweetRepository.save(entity);
        return tweetMapper.toResponseDto(saved);
    }

    @Cacheable(value = "tweets", key = "#id")
    @Transactional(readOnly = true)
    public TweetResponseTo getById(Long id) {
        Tweet entity = tweetRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Tweet with id " + id + " not found", 40405));
        return tweetMapper.toResponseDto(entity);
    }

    @Cacheable(value = "tweets", key = "'all'")
    @Transactional(readOnly = true)
    public List<TweetResponseTo> getAll() {
        return tweetRepository.findAll().stream()
                .map(tweetMapper::toResponseDto)
                .collect(Collectors.toList());
    }

    @Caching(put = @CachePut(value = "tweets", key = "#result.id"),
            evict = @CacheEvict(value = "tweets", key = "'all'"))
    public TweetResponseTo update(TweetRequestTo dto) {
        Tweet existing = tweetRepository.findById(dto.getId())
                .orElseThrow(() -> new ResourceNotFoundException("Tweet with id " + dto.getId() + " not found", 40406));
        if (!editorRepository.existsById(dto.getEditorId())) {
            throw new ResourceNotFoundException("Editor with id " + dto.getEditorId() + " not found", 40407);
        }
        if (tweetRepository.existsByTitleAndIdNot(dto.getTitle(), dto.getId())) {
            throw new DuplicateException("Tweet with title '" + dto.getTitle() + "' already exists", 40304);
        }
        LocalDateTime originalCreated = existing.getCreated();
        tweetMapper.updateEntityFromDto(dto, existing);
        existing.setCreated(originalCreated);
        existing.setModified(LocalDateTime.now());
        Tweet updated = tweetRepository.save(existing);
        return tweetMapper.toResponseDto(updated);
    }

    @Caching(evict = {
            @CacheEvict(value = "tweets", key = "#id"),
            @CacheEvict(value = "tweets", key = "'all'")
    })
    public void deleteById(Long id) {
        if (!tweetRepository.existsById(id)) {
            throw new ResourceNotFoundException("Tweet with id " + id + " not found", 40408);
        }
        tweetRepository.deleteById(id);
    }
}

File: 351002/Abrashin/publisher/src/main/kotlin/com/example/distcomp/service/TweetService.kt
package com.example.distcomp.service

import com.example.distcomp.cache.CacheKeys
import com.example.distcomp.cache.CacheNames
import com.example.distcomp.cache.CacheSupport
import com.example.distcomp.dto.request.TweetRequestTo
import com.example.distcomp.dto.response.CreatorResponseTo
import com.example.distcomp.dto.response.NoteResponseTo
import com.example.distcomp.dto.response.StickerResponseTo
import com.example.distcomp.dto.response.TweetResponseTo
import com.example.distcomp.exception.ConflictException
import com.example.distcomp.exception.NotFoundException
import com.example.distcomp.mapper.CreatorMapper
import com.example.distcomp.mapper.StickerMapper
import com.example.distcomp.repository.CreatorRepository
import com.example.distcomp.repository.StickerRepository
import com.example.distcomp.repository.TweetRepository
import com.example.distcomp.mapper.TweetMapper
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Sort
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDateTime

@Service
class TweetService(
    private val repository: TweetRepository,
    private val mapper: TweetMapper,
    private val creatorRepository: CreatorRepository,
    private val stickerRepository: StickerRepository,
    private val noteService: NoteService,
    private val creatorMapper: CreatorMapper,
    private val stickerMapper: StickerMapper,
    private val cacheSupport: CacheSupport
) {
    @Transactional
    fun create(request: TweetRequestTo): TweetResponseTo {
        val creatorId = request.creatorId ?: throw ConflictException("Creator ID is required")
        if (!creatorRepository.existsById(creatorId)) throw NotFoundException("Creator with id $creatorId not found")

        request.title?.let { title ->
            if (repository.existsByCreatorIdAndTitle(creatorId, title)) {
                throw ConflictException("Tweet with title $title already exists for creator $creatorId")
            }
        }

        val entity = mapper.toEntity(request)
        entity.creatorId = creatorId

        entity.stickers = request.stickers?.map { name ->
            stickerRepository.findByName(name) ?: stickerRepository.save(com.example.distcomp.model.Sticker(name = name))
        } ?: emptyList()

        val now = LocalDateTime.now()
        entity.created = now
        entity.modified = now
        val saved = repository.save(entity)
        return mapper.toResponse(saved).also { response ->
            response.id?.let { cacheSupport.put(CacheNames.TWEETS_BY_ID, it, response) }
            cacheSupport.clear(CacheNames.TWEETS_PAGE)
            cacheSupport.clear(CacheNames.STICKERS_PAGE)
        }
    }

    fun getById(id: Long): TweetResponseTo =
        cacheSupport.getOrPut(CacheNames.TWEETS_BY_ID, id) {
            val entity = repository.findById(id) ?: throw NotFoundException("Tweet with id $id not found")
            mapper.toResponse(entity)
        }

    fun getAll(page: Int, size: Int, sort: Array<String>): List<TweetResponseTo> =
        cacheSupport.getOrPut(CacheNames.TWEETS_PAGE, CacheKeys.page(page, size, sort)) {
            val sortOrder = if (sort.size >= 2) {
                Sort.by(Sort.Direction.fromString(sort[1]), sort[0])
            } else if (sort.isNotEmpty()) {
                Sort.by(sort[0])
            } else {
                Sort.unsorted()
            }
            val pageable = PageRequest.of(page, size, sortOrder)
            repository.findAll(pageable).content.map { mapper.toResponse(it) }
        }

    @Transactional
    fun patch(id: Long, request: TweetRequestTo): TweetResponseTo {
        val existing = repository.findById(id) ?: throw NotFoundException("Tweet with id $id not found")

        request.creatorId?.let { newCreatorId ->
            if (!creatorRepository.existsById(newCreatorId)) throw NotFoundException("Creator with id $newCreatorId not found")
            existing.creatorId = newCreatorId
        }

        val currentCreatorId = existing.creatorId ?: throw ConflictException("Existing tweet has no creator")
        val newTitle = request.title ?: existing.title
        
        if (newTitle != null) {
            val duplicate = repository.existsByCreatorIdAndTitle(currentCreatorId, newTitle)
            if (duplicate && (newTitle != existing.title || request.creatorId != null)) {
                if (newTitle != existing.title || (request.creatorId != null && request.creatorId != existing.creatorId)) {
                    throw ConflictException("Tweet with title $newTitle already exists for creator $currentCreatorId")
                }
            }
        }

        request.title?.let { existing.title = it }
        request.content?.let { existing.content = it }
        request.stickers?.let { names ->
            existing.stickers = names.map { name ->
                stickerRepository.findByName(name) ?: stickerRepository.save(com.example.distcomp.model.Sticker(name = name))
            }
        }

        existing.modified = LocalDateTime.now()
        val saved = repository.save(existing)
        return mapper.toResponse(saved).also { response ->
            cacheSupport.put(CacheNames.TWEETS_BY_ID, id, response)
            cacheSupport.clear(CacheNames.TWEETS_PAGE)
            cacheSupport.evict(CacheNames.TWEET_CREATORS, id)
            cacheSupport.evict(CacheNames.TWEET_STICKERS, id)
            cacheSupport.clear(CacheNames.STICKERS_PAGE)
        }
    }

    fun getCreatorByTweetId(id: Long): CreatorResponseTo =
        cacheSupport.getOrPut(CacheNames.TWEET_CREATORS, id) {
            val tweet = repository.findById(id) ?: throw NotFoundException("Tweet with id $id not found")
            val creatorId = tweet.creatorId ?: throw NotFoundException("Creator for tweet $id not found")
            val creator = creatorRepository.findById(creatorId)
                ?: throw NotFoundException("Creator with id $creatorId not found")
            creatorMapper.toResponse(creator)
        }

    fun getStickersByTweetId(id: Long): List<StickerResponseTo> =
        cacheSupport.getOrPut(CacheNames.TWEET_STICKERS, id) {
            val tweet = repository.findById(id) ?: throw NotFoundException("Tweet with id $id not found")
            tweet.stickers.map { stickerMapper.toResponse(it) }
        }

    fun getNotesByTweetId(id: Long): List<NoteResponseTo> {
        if (!repository.existsById(id)) throw NotFoundException("Tweet with id $id not found")
        return noteService.getNotesByTweetId(id)
    }

    fun delete(id: Long) {
        if (!repository.deleteById(id)) {
            throw NotFoundException("Tweet with id $id not found")
        }
        cacheSupport.evict(CacheNames.TWEETS_BY_ID, id)
        cacheSupport.clear(CacheNames.TWEETS_PAGE)
        cacheSupport.evict(CacheNames.TWEET_CREATORS, id)
        cacheSupport.evict(CacheNames.TWEET_STICKERS, id)
        cacheSupport.evict(CacheNames.NOTES_BY_TWEET_ID, id)
    }
}
